sbt-dotty/src/dotty/tools/sbtplugin/DottyIDEPlugin.scala

```diff
@@ -27,20 +27,99 @@ object DottyIDEPlugin extends AutoPlugin {
     buf.toList
   }
 
-  private def inAllDottyConfigurations[A](key: TaskKey[A], state: State): Task[Seq[A]] = {
-    val struct = Project.structure(state)
-    val settings = struct.data
-    struct.allProjectRefs.flatMap { projRef =>
-      val project = Project.getProjectForReference(projRef, struct).get
+  private def isDottyVersion(version: String) =
+    version.startsWith("0.")
+
+
+  /** Return a new state derived from `state` such that scalaVersion returns `newScalaVersion` in all
+   *  projects in `projRefs` (`state` is returned if no setting needed to be updated).
+   */
+  private def updateScalaVersion(state: State, projRefs: Seq[ProjectRef], newScalaVersion: String): State = {
+    val extracted = Project.extract(state)
+    val settings = extracted.structure.data
+
+    if (projRefs.forall(projRef => scalaVersion.in(projRef).get(settings).get == newScalaVersion))
+      state
+    else {
+      def matchingSetting(setting: Setting[_]) =
+        setting.key.key == scalaVersion.key &&
+        setting.key.scope.project.fold(ref => projRefs.contains(ref), ifGlobal = true, ifThis = true)
+
+      val newSettings = extracted.session.mergeSettings.collect {
+        case setting if matchingSetting(setting) =>
+          scalaVersion in setting.key.scope := newScalaVersion
+      }
+      val newSession = extracted.session.appendRaw(newSettings)
+      BuiltinCommands.reapply(newSession, extracted.structure, state)
+    }
+  }
+
+  /** Setup to run in all dotty projects.
+   *  Return a triplet of:
+   *  (1) A version of dotty
+   *  (2) A list of dotty projects
+   *  (3) A state where `scalaVersion` is set to (1) in all projects in (2)
+   */
+  private def dottySetup(state: State): (String, Seq[ProjectRef], State) = {
+    val structure = Project.structure(state)
+    val settings = structure.data
+
+    val (dottyVersions, dottyProjRefs) =
+      structure.allProjectRefs.flatMap { projRef =>
+        val version = scalaVersion.in(projRef).get(settings).get
+        if (isDottyVersion(version))
+          Some((version, projRef))
+        else
+          crossScalaVersions.in(projRef).get(settings).get.filter(isDottyVersion).sorted.lastOption match {
```

Comment:
Yeah I had a FIXME before that we need a real ordering for version numbers, will add it back.